src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -115,19 +117,139 @@ public sealed override object ReadInline(object item, IEdmTypeReference edmType,
+                        }
+                    }
+                    else
+                    {
+                        ODataDeltaLinkWrapper deltaLinkWrapper = odataItemBase as ODataDeltaLinkWrapper;
+
+                        Contract.Assert(deltaLinkWrapper != null, "ODataDeltaLinkWrapper should not be null");
+
+                        yield return DeserializeDeltaLink(readContext, deltaLinkWrapper, resourceSet.ResourceSetBase.TypeName);                      
+                    }
+                    
+                }
+            }            
+        }
+
+
+        private static IEdmDeltaLinkBase DeserializeDeltaLink(ODataDeserializerContext readContext, ODataDeltaLinkWrapper deltaLinkWrapper, string typeName)
+        {
+            ODataDeltaLinkBase deltalink = deltaLinkWrapper.DeltaLink;
+
+            if (deltalink == null)
+            {
+                throw new ODataException("Deleted link not present");
+            }            
+
+            IEdmModel model = readContext.Model;
+
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            IEdmStructuredType actualType = model.FindType(typeName) as IEdmStructuredType;
+            if (actualType == null)
+            {
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
+            }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            if (actualType.IsAbstract)
             {
-                yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, deltalink.TypeAnnotation.TypeName);
+                throw new ODataException(message);
             }
+
+            IEdmEntityType actualEntityType = actualType as IEdmEntityType;
+
+            IEdmDeltaLinkBase edmDeltaLink;
+            ODataDeltaDeletedLink deletedLink = deltaLinkWrapper.DeltaLink as ODataDeltaDeletedLink;
+            if (deletedLink != null)
+            {
+                edmDeltaLink = new EdmDeltaDeletedLink(actualEntityType);
+            }
+            else
+            {
+                edmDeltaLink = new EdmDeltaLink(actualEntityType);
+            }
+
+            edmDeltaLink.Source = deltalink.Source;
+            edmDeltaLink.Target = deltalink.Target;
+            edmDeltaLink.Relationship = deltalink.Relationship;
+
+            return edmDeltaLink;
+        }
+
+        private static EdmDeltaDeletedEntityObject DeSerializeDeletedEntity(ODataDeserializerContext readContext, ODataDeletedResource deletedResource)
+        {
+            if (deletedResource == null)
+            {
+                throw new ODataException("Deleted resource not present");
+            }
```

Comment:
this should be an assert -- you only called this because you determined that deletedResource was non-null.